File: netball.jl
using Lux
import Random
import Optimisers
import Zygote
using ComponentArrays
using DiffEqFlux
using CSV
using DataFrames
using DifferentialEquations
using Plots
# Load data
data = CSV.read("./netball.csv", DataFrame)
dudt_model = Lux.Chain(
  WrappedFunction(u -> u[3:4]),
  Lux.Parallel(vcat,
# extract (ẋ, ẏ) from input (x, y, ẋ, ẏ)
# output is concatenation of:
#   inputs (ẋ, ẏ)
#   neural network a(ẋ, ẏ)
NoOpLayer(),
Lux.Chain(
      Lux.Dense(2 => 16, tanh),
      Lux.Dense(16 => 16, tanh),
      Lux.Dense(16 => 16, tanh),
      Lux.Dense(16 => 2)
) )
)
ps, st = Lux.setup(Random.default_rng(), dudt_model)
ps = ComponentArray(ps)

u0 = [data.x[1], data.y[1], data.v[1]*cosd(data.theta[1]), data.v[1]*sind(data.theta[1])]
xy_training = collect(Array(data[!,[:x,:y]])')

function L(ps)
  xy_predicted = Array(prob(u0, ps, st)[1])[1:2,:]
  sum(abs2, xy_predicted .- xy_training)
end

opt = Optimisers.setup(Optimisers.Adam(0.01f0), ps)
for epoch ∈ 1:1000#5000
  if rem(epoch, 10) == 0
    println("epoch $(epoch): Loss = $(L(ps))")
  end
  gs = Zygote.gradient(L, ps)[1]            # compute gradient wrt ps
  opt, ps = Optimisers.update(opt, ps, gs)  # update ps using the gradient
end

# NeuralODE
println("Execution time for Julia(DiffEqFlux.NeuralODE) implememt Neural ODEs:")
@time begin
prob = NeuralODE(dudt_model, (data.t[1], data.t[end]), Tsit5(); saveat=data.t)
end


xy_predicted=Array(prob(u0, ps, st)[1])[1:2,:]
x_predicted = xy_predicted[1, :]
y_predicted = xy_predicted[2, :]

# Plot the trajectories
plot(
    x_predicted, y_predicted, label="Predicted trajectory",
    xlabel="x_predicted", ylabel="y_predicted", title="Netball trajectory (Implementing with Julia)", legend=:topleft
)
plot!(
    data.x, data.y, label="True trajectory",
    xlabel="x", ylabel="y"
)

File: netball.py
import numpy as np
import pandas as pd
import torch
import torch.optim as optim
import torch.nn as nn
from torchdiffeq import odeint
import matplotlib.pyplot as plt
import time
import sys
import gc

# Load data
data = pd.read_csv("./netball.csv")
x0, y0, v0, theta0 = data.loc[0, ['x', 'y', 'v', 'theta']]

# Define neural network model
class dudt_model(nn.Module):
    def __init__(self):
        super(dudt_model, self).__init__()
        self.net = nn.Sequential(
            nn.Linear(2, 16),
            nn.Tanh(),
            nn.Linear(16, 16),
            nn.Tanh(),
            nn.Linear(16, 16),
            nn.Tanh(),
            nn.Linear(16, 2)
        )

    def forward(self, t, y):
        return self.net(y)

# Define loss function
def loss_fn(pred_traj, true_traj):
    return torch.sum((pred_traj - true_traj) ** 2)

# Define initial conditions
u0 = torch.tensor([x0, y0, v0 * np.cos(np.radians(theta0)), v0 * np.sin(np.radians(theta0))], dtype=torch.float32)

# Prepare true trajectory
true_traj = torch.tensor(data[['x', 'y']].values, dtype=torch.float32)

# Define model
model = dudt_model()

# Define optimizer
optimizer = optim.Adam(model.parameters(), lr=0.01)

# Training loop
epochs = 1000
for epoch in range(epochs):
    optimizer.zero_grad()
    pred_traj = odeint(model, u0[:2], torch.tensor(data['t'].values, dtype=torch.float32))
    loss = loss_fn(pred_traj, true_traj)
    loss.backward()
    optimizer.step()
    if epoch % 10 == 0:
        print(f"Epoch {epoch}: Loss = {loss.item()}")
print(f"Epoch {epochs}: Loss = {loss.item()}")
# Predict trajectory using Neural ODEs
# Get current memory usage
start_memory = gc.get_objects()
# Get current time
start_time = time.time()
pred_traj = odeint(model, u0[:2], torch.tensor(data['t'].values, dtype=torch.float32))
end_time = time.time()
end_memory = gc.get_objects()
execution_time = end_time - start_time
print("Execution time for Python(torchdiffeq.odeint) implememt Neural ODEs:\n", execution_time, "seconds")
# Calculate the increment of memory allocation
allocations = len(end_memory) - len(start_memory)
# Calculate the size of memory allocation
memory_allocation = sum(sys.getsizeof(obj) for obj in end_memory) - sum(sys.getsizeof(obj) for obj in start_memory)
print(f"{allocations} allocations: {memory_allocation / 1024:.3f} KiB")

# Plot trajectories
plt.plot(pred_traj[:, 0].detach().numpy(), pred_traj[:, 1].detach().numpy(), label="Predicted trajectory")
plt.plot(data['x'], data['y'], label="True trajectory")
plt.xlabel('x')
plt.ylabel('y')
plt.title('Netball trajectory (Implementing with Python)')
plt.legend()
plt.show()